scripts/handleTriggerDoors.js

```diff
@@ -0,0 +1,72 @@
+/**
+ * Gets the comma delimited keys specified in the "assignedDoors" property belonging to a Tiled object and returns them in an array
+ * For example, "hello, world, foo, bar" would return ["hello", "world", "foo", "bar"]
+ * @param {object} entity - A Tiled object
+ * @returns An array of all the doors assigned to the specified Tiled object.
+ */
+function getEntityAssignedDoors(entity) {
+  return entity.assignedDoors
+    .replace(/\s/g, "")
+    .split(",")
+    .filter((assignedDoor) => !!assignedDoor);
+}
+
+function runDoorAction(world, assignedDoors, action) {
+  world.forEachEntities(
+    ({ instance }) => {
+      return assignedDoors.some((doorKey) => instance.key === doorKey);
+    },
+    (door) => {
+      const requiredCurrentState = action === "open" ? "closed" : "open";
+
+      // This handles players moving in and out of the door trigger by enqueuing a pending state
+      // if the trigger is entered/exited while the door is changing states. For example,
+      // if the door is opening and the player leaves the trigger, the "close" action will
+      // be queued up to run once the door is finished opening.
+      if (door.state.fsm.currentState !== requiredCurrentState) {
```

Comment:
What doors is this applying to? 

I see the FSM so I'm guessing they're a custom object using the [openCloseFsm](https://github.com/TwilioQuest/twilioquest-base/blob/main/scripts/initOpenCloseFsm.js)?

If this code is patching flaws in that FSM, we should make those changes in that script so it benefits other doors as well. 

The old FSM code should handle a transition to the opening state while in the closing state and vice versa. If it doesn't handle correctly, i'd like to fix it there.

Reply:
> What doors is this applying to?

Technically speaking "door" in this context refers to any object that has a key which exists within `assignedDoors` (I may not have fully understood the question haha)

> I see the FSM so I'm guessing they're a custom object using the [openCloseFsm](https://github.com/TwilioQuest/twilioquest-base/blob/main/scripts/initOpenCloseFsm.js)?

Exactly! I was really only concerned with the [door-hopper](https://github.com/TwilioQuest/twilioquest-base/blob/main/objects/door-hopper/config.js#:~:text=const%20initOpenCloseFsm%20%3D%20require\(%22../../scripts/initOpenCloseFsm%22\)%3B), [door-lovelace](https://github.com/TwilioQuest/twilioquest-base/blob/main/objects/door-lovelace/config.js#:~:text=Raw,initOpenCloseFsm%22\)%3B), [door-neumann](https://github.com/TwilioQuest/twilioquest-base/blob/main/objects/door-neumann/config.js#:~:text=const%20initOpenCloseFsm%20%3D%20require\(%22../../scripts/initOpenCloseFsm%22\)%3B), and [door-turing](https://github.com/TwilioQuest/twilioquest-base/blob/main/objects/door-turing/config.js#:~:text=const%20initOpenCloseFsm%20%3D%20require\(%22../../scripts/initOpenCloseFsm%22\)%3B) doors when writing the code, but any custom object that uses the openCloseFsm would work

> If this code is patching flaws in that FSM, we should make those changes in that script so it benefits other doors as well.

Agreed!

> The old FSM code should handle a transition to the opening state while in the closing state and vice versa. If it doesn't handle correctly, i'd like to fix it there.

It's not necessarily that the fsm handles transitions improperly, it's more so that the door animations start on either a "fully closed" frame or "fully opened" frame. So if you send an "open" action while a door is in the "closing" state, it'll handle the transition as we would expect, but since the door was on a frame in-between "fully closed" and "fully opened", there's a noticeable hiccup as it snaps directly to one of the two starting frames.

I'm happy to whip together an issue for this in the twilioquest repo, though! Let me know what you think and I'll hop on it!

Reply:
Ah, yes, I have seen that issue in the past. I believe the fog owl doors handle this with the `smoothTranstionBetweenOpeningAndClosing` property in its config.

I see that it is set to `false` for the [api academy doors](https://github.com/TwilioQuest/twilioquest-base/blob/main/objects/door-hopper/config.js#L37). So that's the part where I'm wondering if this property was overlooked or if there's a different issue when the property is set to `true` that this code fixes?

Reply:
> Ah, yes, I have seen that issue in the past. I believe the fog owl doors handle this with the `smoothTranstionBetweenOpeningAndClosing` property in its config.
> 
> I see that it is set to `false` for the [api academy doors](https://github.com/TwilioQuest/twilioquest-base/blob/main/objects/door-hopper/config.js#L37). So that's the part where I'm wondering if this property was overlooked or if there's a different issue when the property is set to `true` that this code fixes?

I can definitely give this a try! I remember seeing the `smoothTransitionBetweenOpeningAndClosing` property, but it completely slipped my mind haha. I'll come back with an update once I've been able to test this!